apis/storage/v1alpha3/bucketpolicy_types.go

```diff
@@ -0,0 +1,197 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// S3BucketPolicyParameters define the desired state of an AWS S3BucketPolicy.
+type S3BucketPolicyParameters struct {
+	// This is the current IAM policy version
+	PolicyVersion string `json:"Version"`
+
+	// This is the policy's optional identifier
+	PolicyID string `json:"Id,omitempty"`
+
+	// This is the list of statement this policy applies
+	PolicyStatement []S3BucketPolicyStatement `json:"Statement"`
+}
+
+//Serialize is the custom marshaller for the S3BucketPolicyParameters
+func (p *S3BucketPolicyParameters) Serialize() (interface{}, error) {
+	m := make(map[string]interface{})
+	m["Version"] = p.PolicyVersion
+	if p.PolicyID != "" {
+		m["Id"] = p.PolicyID
+	}
+	slc := make([]interface{}, len(p.PolicyStatement))
+	for i, v := range p.PolicyStatement {
+		msg, err := v.Serialize()
+		if err != nil {
+			return nil, err
+		}
+		slc[i] = msg
+	}
+	m["Statement"] = slc
+	//dat, err := json.Marshal(m)
+	//if err != nil {
+	//	return nil, err
+	//}
+	//str := string(dat)
+	return m, nil
+}
+
+// S3BucketPolicyStatement defines an individual statement within the
+// S3BucketPolicyBody
+type S3BucketPolicyStatement struct {
+	// Optional identifier for this statement, must be unique within the
+	// policy if provided.
+	StatementID string `json:"Sid,omitempty"`
+
+	// The effect is required and specifies whether the statement results
+	// in an allow or an explicit deny. Valid values for Effect are Allow and Deny.
+	Effect string `json:"Effect"`
+
+	// Used with the S3 policy to specify the principal that is allowed
+	// or denied access to a resource.
+	Principal *S3BucketPrincipal `json:"Principal,omitempty"`
+
+	// Each element of the PolicyAction array describes describes the specific
+	// action or actions that will be allowed or denied with this PolicyStatement.
+	PolicyAction []string `json:"Action"`
+
+	// This flag indicates that this policy should apply to the IAMUsername
+	// that was either passed in or created for this bucket.
+	ApplyToIAMUser bool `json:"EffectIAMUser,omitempty"`
+
+	// The paths on which this resource will apply
+	ResourcePath []string `json:"Resource"`
+}
+
+//Serialize is the custom marshaller for the S3BucketPolicyStatement
+func (p *S3BucketPolicyStatement) Serialize() (interface{}, error) {
```

Comment:
Is there a reason to implement `Serialize()` as opposed to just satisfying `json.Marshaler` interface by implementing custom `MarshalJSON()`?

Reply:
The reason I implemented the Serialize() function as opposed to the custom Marshaller/Unmarshaller is that the custom marshaling interfered with certain functionality in the crossplane-runtime and the Kubernetes client. This resulted in unexpected behavior and run-time errors. Additionally, AWS changed the shape of the policy after it is sent (i.e., if I sent a policy, then I query the API for that policy again, the shape of the queried policy could be different). To implement the update functionality, I needed to be able to consistency compare the current policy and the policy from the API.